core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java

```diff
@@ -0,0 +1,216 @@
+        final HttpResponse response = action.apply(req);
+        final HttpResponseDuplicator duplicator = new HttpResponseDuplicator(response);
+        retryStrategy().shouldRetry(req, duplicator.duplicateStream())
+                       .handle(voidFunction((retry, unused) -> {
+            if (retry != null && retry) {
+                long millisAfter = useRetryAfter ? getRetryAfterMillis(duplicator.duplicateStream()) : -1;
+                duplicator.close();
+                retry0(currentAttemptNo, backoff, ctx, req, action,
+                       rootResponse, RetryException.get(), millisAfter);
+            } else {
+                duplicator.duplicateStream().aggregate().handle(voidFunction((aRes, cause) -> {
+                final Throwable exception;
+                try {
+                    if (cause != null) {
+                        if (!retryStrategy().shouldRetry(req, cause)) {
+                            rootResponse.close(cause);
+                            return;
+                        }
+                        exception = cause;
+                    } else {
+                        rootResponse.respond(aRes);
+                        return;
+                    }
+                } finally {
+                    duplicator.close();
+                }
+                retry0(currentAttemptNo, backoff, ctx, req, action,
+                       rootResponse, exception, -1);
+                }));
+            }
+
+        }));
+    }
+
+    private long getRetryAfterMillis(HttpResponse res) {
+        final HttpHeaders headers = getHttpHeaders(res);
+        long millisAfter = -1;
+        String value = headers.get(HttpHeaderNames.RETRY_AFTER);
+        if (value != null) {
+            if (isSecond(value)) {
+                millisAfter = TimeUnit.SECONDS.toMillis(Integer.parseInt(value));
+            } else {
+                try {
+                    final ZonedDateTime later =
+                            ZonedDateTime.from(DateTimeFormatter.RFC_1123_DATE_TIME.parse(value));
+                    final ZonedDateTime current = ZonedDateTime.now(later.getZone());
+                    millisAfter = ChronoUnit.MILLIS.between(current, later);
+                } catch (Exception ignored) {
+                    // Neither a http date nor a second.
+                }
+            }
+        }
+        return millisAfter;
+    }
+
+    private HttpHeaders getHttpHeaders(HttpResponse res) {
+        final CompletableFuture<HttpHeaders> future = new CompletableFuture<>();
+        final HeaderSubscriber subscriber = new HeaderSubscriber(future);
+        res.closeFuture().whenComplete(subscriber);
+        res.subscribe(subscriber);
+        return future.join(); // Neither blocks here nor throws an exception because it already has headers.
+    }
+
+    private boolean isSecond(String s) {
+        return s.matches("^-?\\d+$");
```

Comment:
That was my first implementation.
I thought that If most of inputs are integers, it will be good using tryParse like you said.
However, if it's not. it will throw newly created NumberFormatException every time I try parsing.
I was not sure which one is better.